src/simplex.pybind.h

```diff
@@ -70,27 +71,116 @@ void add_simplex_module(py::module& root) {
+            // for direct NumPy array creation. Therefore, I use `std::vector<char>`
+            // to ensure compatibility with `py::array`.
+            std::vector<char> result(self.num_observables, 0);
             for (size_t ei : self.predicted_errors_buffer) {
               for (int obs_index : self.errors[ei].symptom.observables) {
-                result[obs_index] = result[obs_index] ^ true;
+                result[obs_index] = result[obs_index] ^ 1;
               }
             }
+
+            return py::array(py::dtype::of<bool>(), result.size(), result.data());
+          },
+          py::arg("syndrome"),
+          py::call_guard<py::scoped_ostream_redirect, py::scoped_estream_redirect>(),
+          R"pbdoc(
+        Decodes a single shot.
+
+        Parameters
+        ----------
+        syndrome : np.ndarray
+            A 1D NumPy array of booleans representing the detection events for a single shot.
+            The length of the array should match the number of detectors in the DEM.
+
+        Returns
+        -------
+        np.ndarray
+            A 1D NumPy array of booleans indicating which observables are flipped.
+            The length of the array matches the number of observables.
+    )pbdoc")
+      .def(
+          "decode_batch",
+          [](SimplexDecoder& self, const py::array_t<bool>& syndromes) {
+            // Check the dimensions of the `syndromes` argument.
+            if (syndromes.ndim() != 2) {
+              throw std::runtime_error("Input syndromes must be a 2D NumPy array.");
+            }
+
+            // Retrieve the number of shots, detectors and the syndrome patterns.
+            auto syndromes_unchecked = syndromes.unchecked<2>();
+            size_t num_shots = syndromes_unchecked.shape(0);
+            size_t num_detectors = syndromes_unchecked.shape(1);
+
+            // Allocate the result array.
+            py::array_t<bool> result({num_shots, self.num_observables});
+            auto result_unchecked = result.mutable_unchecked<2>();
+
+            // Process and decode each shot.
+            for (size_t i = 0; i < num_shots; ++i) {
+              std::vector<uint64_t> detections;
+              for (size_t j = 0; j < num_detectors; ++j) {
+                if (syndromes_unchecked(i, j)) {
+                  detections.push_back(j);
+                }
+              }
+              self.decode(detections);
+
+              // Collect results for the current shot being decoded.
+              std::vector<char> shot_result(self.num_observables, 0);
+              for (size_t ei : self.predicted_errors_buffer) {
+                for (int obs_index : self.errors[ei].symptom.observables) {
+                  shot_result[obs_index] ^= 1;
+                }
+              }
+
+              // Copy the result into the pre-allocated array.
+              for (size_t k = 0; k < self.num_observables; ++k) {
+                result_unchecked(i, k) = shot_result[k];
+              }
```

Comment:
Similar to my comment below, it might be simpler to do something like this which avoids the copying:

```suggestion
              for (size_t ei : self.predicted_errors_buffer) {
                for (int obs_index : self.errors[ei].symptom.observables) {
                  result_unchecked(i, obs_index) ^= 1;
                }
              }
```

Reply:
@oscarhiggott Thank you. Done.